ipv8/configuration_pydantic.py

```diff
@@ -0,0 +1,357 @@
+    interface: str = default["interfaces"][0]["interface"]
+    ip: str = default["interfaces"][0]["ip"]
+    port: int = default["interfaces"][0]["port"]
+
+
+class Logger(BaseConfigurationClass):
+    level: str = default["logger"]["level"]
+
+    @validator("level", always=True)
+    @classmethod
+    def validate_level(cls, v):
+        allowed = ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOTSET']
+        assert v in allowed, f"Illegal log level {v} specified! Should be one of {allowed}!"
+        return v
+
+
+class Key(BaseConfigurationClass, AbstractModel):
+    alias: str = default["keys"][0]["alias"]
+
+
+class PreloadedKey(Key):
+    bin: str
+
+
+def generate_ephemeral_key():
+    return base64.b64encode(default_eccrypto.generate_key("curve25519").key_to_bin()).decode()
+
+
+class EphemeralKey(Key):
+    bin: str = Field(frozen=True, default_factory=generate_ephemeral_key)
+
+
+class FileKey(Key):
+    generation: str = default["keys"][0]["generation"]
+    file: str = default["keys"][0]["file"]
+
+
+class Walker(BaseConfigurationClass, AbstractModel):
+    strategy: str = Field(...)
+    peers: int = 20
+    init: dict = Field(default_factory=dict)
+
+    @validator("peers", always=True)
+    @classmethod
+    def validate_peer_count(cls, v):
+        assert v >= 0 or v == -1, f"A walker's peer count must be >= 0 or set to -1 (infinite), got: {v}!"
+        return v
+
+
+class RandomWalk(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "RandomWalk"
+        if "init" not in kwargs:
+            kwargs["init"] = {"timeout": 3.0}
+        super().__init__(**kwargs)
+
+
+class EdgeWalk(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "EdgeWalk"
+        super().__init__(**kwargs)
+
+
+class RandomChurn(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "RandomChurn"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        if "init" not in kwargs:
+            kwargs["init"] = {'sample_size': 8, 'ping_interval': 10.0, 'inactive_time': 27.5, 'drop_time': 57.5}
+        super().__init__(**kwargs)
+
+
+class PeriodicSimilarity(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "PeriodicSimilarity"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        super().__init__(**kwargs)
+
+
+class PingChurn(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "PingChurn"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        super().__init__(**kwargs)
+
+
+class Overlay(BaseConfigurationClass):
+    cls: str = Field(alias='class')
+    key: str = default["keys"][0]["alias"]
+    walkers: List[Walker] = [RandomWalk()]
+    bootstrappers: List[Bootstrapper] = [DispersyBootstrapper(init=DispersyBootstrapperInit())]
+    initialize: dict = Field(default_factory=dict)
+    on_start: List[tuple] = []
+
+
+class IPv8Configuration(BaseConfigurationClass):
```

Comment:
Thanks! The problem is a bit worse though: this is already the third configuration option (with `dict` and `ConfigBuilder` being the other two). I agree that one option should be leading to avoid duplication and for this IPv8 uses the `dict` as the leading config. The reasoning for this is that _(a)_ `dict` requires no external dependencies, allowing all other options to be optional (and less hard dependencies is better for just about every metric), _(b)_ most configuration libraries can convert to a `dict` relatively easily, _(c)_ pulling from the common source `dict` does not require one optional configuration library to depend on importing another (now non-optional) configuration library that is otherwise unused.

The IPv8 ethos is to support everything for everyone as much as possible. IPv8 does not care about who (or what) calls it and whether they think XML is superior to pydantic or JSON. It should all _just work_, no matter what your preference may be.

Reply:
> I agree that one option should be leading to avoid duplication and for this IPv8 uses the dict as the leading config.

Not sure this is true. `pydantic` model could not be created from "leading" dict. That means it does not avoid duplication but produces it.

As far as I understand the code, the description of `ConfigBuilder`'s fields also could not be created from the `dict`, which doesn't make the `dict` a single source of the configuration's description.

If for changing/adding a field to the configuration, you should add it in three places, then it shows there is a problem in the internal configs implementation. 
This implementation adds one more possibility for future developers of ipv8 to make a mistake while changing configs.

This sentence is unclear to me:
> The IPv8 ethos is to support everything for everyone as much as possible. IPv8 does not care about who (or what) calls it and whether they think XML is superior to pydantic or JSON. It should all just work, no matter what your preference may be.

As a library, `IPv8` should provide a clear API for users, without keeping in mind any ethos. That means it should be clear for users which options `IPv8` has for pushing a configuration to it.

However, this reasoning is irrelevant to the problem described above, as the problem is with the internal config organization rather than the user API.

Reply:
> As far as I understand the code, the description of ConfigBuilder's fields also could not be created from the dict, which doesn't make the dict a single source of the configuration's description.

This is incorrect. In fact, the `ConfigBuilder` IS just that same dict, there is no other state:

https://github.com/Tribler/py-ipv8/blob/ce53a318c6e36846f19dd92d39b9014c90516255/ipv8/configuration.py#L193-L195

> This implementation adds one more possibility for future developers of ipv8 to make a mistake while changing configs.

This is true and it is the reason I never added support for pydantic before. However, pydantic appears to be popular.

> As a library, IPv8 should provide a clear API for users, without keeping in mind any ethos. That means it should be clear for users which options IPv8 has for pushing a configuration to it.

Ethos is extremely important. Different domains have very different APIs. For example, High-Performance Computing libraries have vastly different API looks/feels/controls/outlooks than networking APIs or HTTP APIs: you would "never" find a REST API when using HPC or OS libaries. The clarity of use comes with the expertise and domain of its users (networking/distributed systems engineers in this case).

I agree with the latter half of your statement, regarding the code showing what options for configuration exist, which is why I explicitly specified the permissible types that can serve as a configuration.

Reply:
> This is incorrect. In fact, the ConfigBuilder IS just that same dict, there is no other state.

Well. Try to rename the field `walker_interval` and you will see that you have to do it at least in three places. In the `dict`, in the `ConfigBuilder` and in the `pydantic` configuration. This example shows that the config definition is duplicated across the codebase.

As a counterargument, you will probably say that the `walker_interval` name would never be changed because the config must be backward compatible. It is true, but it is not the counterargument to the statement that "config definition is duplicated across the codebase".

As a less simple, but more realistic example: imagine a developer that wants to add a new field to the config. They should change at least three different places in IPv8 codebase:  the `dict`, in the `ConfigBuilder` and in the `pydantic`.

> This implementation adds one more possibility for future developers of ipv8 to make a mistake while changing configs.
This is true and it is the reason I never added support for pydantic before. However, pydantic appears to be popular.

I completely agree with adding support for pydantic. I'm only questioning the way it is implemented right now.

As an approach that could decrease code duplication, I suggest making `pydantic` the "main" place when the config is defined. It is not hard and it is backward compatible.

Reply:
> As a counterargument, you will probably say that the walker_interval name would never be changed because the config must be backward compatible.

No I wouldn't. I'd argue that `walker_interval` is the setting name used throughout the entire project. The names of (most of) the settings directly relate to the parameters of classes in the project. Reducing duplication in hard-coded _values_ is good. However, reducing duplication in the _keys_ of the configuration is impossible. You cannot have two configurations set a different key for the same configurable class. I can provide examples if you like.

> It is not hard and it is backward compatible.

It makes `pydantic` non-optional. I can provide crash logs if you give me an implementation.

Reply:
"Dependency hell" is something that goes back decades. People have spent years of development time on refactoring to newer and hipper libraries (e.g., https://github.com/Tribler/tribler/issues/68). Better alternatives get discovered, support gets dropped, and external libraries turn out to not work quite as well on all operating systems as years go by. Keeping external libraries optional, if they can reasonably be optional, saves a lot of work in the long run.

Even if pydantic is a safe bet now, I don't think it's unreasonable to expect pydantic to be dethroned in the future, by [Gin](https://github.com/google/gin-config) for example. Future generations will curse our names if we go all-in when we don't have to. Therefore, it seems prudent to keep pydantic optional.

Reply:
Your concerns are understandable and valid. 
But I doubt that the `pydantic` dependency is a correct example of the "dependency hell" phenomenon.

> Future generations will curse our names if we go all-in when we don't have to.

Agree, but I don't think this is correct in regard to `pydantic`. Even if Gin dethrones it in the future, this will not necessarily lead to the need of rewriting the codebase. 
However, even if you have to replace  `pydantic` with something else, it is ok as long as `pydantic` configuration remains clean and simple. This is one more point to remove unnecessary and complex `AbstactModel`. 

Future generations will curse our names if they should rewrite the unreasonably complicated configuration.

Reply:
I do agree that code should not be unreasonably complex.

My test for dependency hell is the following question: "Can the dependency reasonably be made optional". The rest of my flow chart is as follows:

 - **No:** restructure the code and use it everywhere.
 - **Yes:** make it optional.

I think our main disagreement here is whether it is reasonable to make pydantic optional. Considering the fact that this is controlled by a single import statement (`import configuration_pydantic`) I don't see any spectacular or unreasonable complexity. I will agree that it makes a singular unit test case need a single if-statement and that is more complex than having no if-statement.

Let's keep the `AbstractModel` discussion in the other thread, or we will derail this one.

Reply:
> I think our main disagreement here is whether it is reasonable to make pydantic optional

Yes, I guess so :)